internal/watcher/watcher_test.go

```diff
@@ -0,0 +1,204 @@
+			t.Fatalf("ioutil.WriteFile: %v", err)
+		}
+
+		file = filepath.Join(subDir, fmt.Sprintf("test_file_%d.swp", i))
+		if err = ioutil.WriteFile(file, []byte(":octopus:"), 0755); err != nil {
+			t.Fatalf("ioutil.WriteFile: %v", err)
+		}
+	}
+
+	files, directories, err := GetAllInDirectory(baseDir, ignoredExtensions)
+	if err != nil {
+		t.Fatalf("GetAllInDirectory: %v", err)
+	}
+
+	t.Run("Count files", func(t *testing.T) {
+		assert.Equal(t, expectedFilesCount, len(files))
+	})
+
+	t.Run("Count directories", func(t *testing.T) {
+		assert.Equal(t, expectedDirCount, len(directories))
+	})
+
+	t.Run("Ignored extensions", func(t *testing.T) {
+		for _, f := range files {
+			if _, ok := ignoredExtensionsMap[filepath.Ext(f)]; ok {
+				assert.Fail(t, "Haven't excluded files with ignored extensions.")
+				return
+			}
+		}
+	})
+}
+
+type mockWatchFuncProvider struct {
+	called      bool
+	calledMutex *sync.Mutex
+}
+
+func (m *mockWatchFuncProvider) WatchFn(path string) error {
+	m.calledMutex.Lock()
+	m.called = true
+	m.calledMutex.Unlock()
+
+	return nil
+}
+
+func TestWatchDirectories(t *testing.T) {
+	mockWatchFuncProvider := &mockWatchFuncProvider{
+		calledMutex: &sync.Mutex{},
+	}
+
+	baseDir, err := ioutil.TempDir("", "watcher_test1")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir: %v", err)
+	}
+
+	defer os.RemoveAll(baseDir)
+
+	file := filepath.Join(baseDir, "test_file")
+	if err = ioutil.WriteFile(file, []byte(":octopus:"), 0755); err != nil {
+		t.Fatalf("ioutil.WriteFile: %v", err)
+	}
+
+	files, directories, err := GetAllInDirectory(baseDir, []string{".swp"})
+	if err != nil {
+		t.Fatalf("GetAllInDirectory: %v", err)
+	}
+
+	done := make(chan struct{})
+
+	go func() {
+		for {
+			<-time.After(100 * time.Millisecond)
+
+			if err := ioutil.WriteFile(files[0], []byte(":octopus: - :octopus:"), 0755); err != nil {
+				t.Logf("ioutil.WriteFile: %v", err)
+			}
+
+			if mockWatchFuncProvider.called {
+				break
+			}
+		}
+
+		close(done)
+	}()
+
+	if err := WatchDirectories(directories, mockWatchFuncProvider, done); err != nil {
+		t.Fatalf("WatchDirectories: %v", err)
+	}
+
+	assert.Equal(t, true, mockWatchFuncProvider.called)
+}
+
+func TestNewDifferential(t *testing.T) {
+	t.Run("Empty struct", func(t *testing.T) {
+		expected := Differential{
+			previousChecksum: make(map[string][16]byte),
+			previousContents: make(map[string][]byte),
+			client:           nil,
+			url:              "",
+			token:            "",
+		}
+
+		actual, err := NewDifferential([]string{}, nil, "", "")
+		if err != nil {
+			t.Fatalf("NewDifferential: %v", err)
+		}
+
+		assert.Equal(t, expected, actual)
+	})
+
+	t.Run("Checksum & contents", func(t *testing.T) {
+		previousChecksum := make(map[string][16]byte)
+		previousContents := make(map[string][]byte)
+
+		baseDir, err := ioutil.TempDir("", "watcher_test1")
+		if err != nil {
+			t.Fatalf("ioutil.TempDir: %v", err)
+		}
+
+		defer os.RemoveAll(baseDir)
+
+		var b []byte
```

Comment:
You can assign this in the loop. It doesn't have to be in this scope, I guess.

Reply:
My mistake